Create new lighttpd pihole-admin.conf for admin interface

gstrauss opened the pull request:
**What does this PR aim to accomplish?:**

lighttpd pihole-admin.conf for admin interface

First steps to simplifying the convoluted historical pi-hole configuration of lighttpd.

**How does this PR accomplish the above?:**

PR creates pihole-admin.conf and update install and uninstall scripts.

---
**By submitting this pull request, I confirm the following:**

1. I have read and understood the [contributors guide](https://docs.pi-hole.net/guides/github/contributing/), as well as this entire template. I understand which branch to base my commits and Pull Requests against.
2. I have commented my proposed changes within the code and I have tested my changes.
3. I am willing to help maintain this change if there are issues with it later.
4. It is compatible with the [EUPL 1.2 license](https://opensource.org/licenses/EUPL-1.1)
5. I have squashed any insignificant commits. ([`git rebase`](http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html))
6. I have checked that another pull request for this purpose does not exist.
7. I have considered, and confirmed that this submission will be valuable to others.
8. I accept that this submission may not be used, and the pull request closed at the will of the maintainer.
9. I give this submission freely, and claim no ownership to its content.

---
- [x] I have read the above and my PR is ready for review. *Check this box to confirm*

jfb-pihole commented:
Is this PR warranted, given that we are moving away from lighttpd in V6?

rdwebdesign commented:
@jfb-pihole 

We are aware of this PR.

PromoFaux commented on advanced/lighttpd.conf.fedora:
```diff
@@ -86,37 +85,4 @@ fastcgi.server = (
-
-# Default expire header
-expire.url = ( "" => "access plus 0 seconds" )
+include "/etc/lighttpd/conf.d/pihole-admin.conf"
```
Is there a reason that this line is added to `lighttpd.conf.fedora`, but not to `lighttpd.conf.debian`?

gstrauss replied:
`lighttpd.conf.debian` already contains `include_shell "find /etc/lighttpd/conf-enabled -name '*.conf' -a ! -name 'letsencrypt.conf' -printf 'include \"%p\"\n' 2>/dev/null"`

gstrauss commented:
I re-pushed (force-pushed) with my commits signed.

gstrauss commented:
I added commit with change from #5069 to set "X-XSS-Protection: 0"

yubiuser reviewed (dismissed):
Please adjust the test suite to include a test for the new file.

dschaper commented:
> I added commit with change from #5069 to set "X-XSS-Protection: 0"

Please revert that. There is another PR open that addresses this and it is documented with the reasons for that change. I prefer to have PRs that address one issue at a time.

gstrauss commented:
>> I added commit with change from https://github.com/pi-hole/pi-hole/pull/5069 to set "X-XSS-Protection: 0"

> Please revert that. There is another PR open that addresses this and it is documented with the reasons for that change. I prefer to have PRs that address one issue at a time.

#5069 (submitted after this PR) causes a merge conflict due to the config moving to a different file.  I incorporated the change in this PR.  "Reverting" the commit -- as you requested -- without taking other action to incorporate the change would end up dropping the change.  I have instead incorporated the change.

gstrauss commented:
> Please adjust the test suite to include a test for the new file.

@yubiuser I had added a test for existence of installed file `pihole-admin.conf`.

Future-looking, I thikn such tests are redundant and unnecessary when there is already a test for checking that the pihole admin page is reachable.

gstrauss commented:
Note: these changes are rebased onto HEAD of development, and it appears that the docker images are failing to spawn to run the distro tests.

PromoFaux commented:
Please rebase on `development`, which should fix the test issues

gstrauss commented:
> Please rebase on development, which should fix the test issues

rebased on development and tests now pass

PromoFaux commented on automated install/basic-install.sh:
```diff
@@ -1412,14 +1412,23 @@ installConfigs() {
+        # Copy the config file to include for pihole admin interface
+        if [[ -d "/etc/lighttpd/conf.d" ]]; then
+            install -D -m 644 -T ${PI_HOLE_LOCAL_REPO}/advanced/pihole-admin.conf /etc/lighttpd/conf.d/pihole-admin.conf
+        elif [[ -d "/etc/lighttpd/conf-available2" ]]; then
```
```suggestion
        elif [[ -d "/etc/lighttpd/conf-available" ]]; then
```

PromoFaux commented on automated install/basic-install.sh:
```diff
@@ -1412,14 +1412,23 @@ installConfigs() {
+        if [[ -d "/etc/lighttpd/conf.d" ]]; then
+            install -D -m 644 -T ${PI_HOLE_LOCAL_REPO}/advanced/pihole-admin.conf /etc/lighttpd/conf.d/pihole-admin.conf
+        elif [[ -d "/etc/lighttpd/conf-available2" ]]; then
+            install -D -m 644 -T ${PI_HOLE_LOCAL_REPO}/advanced/pihole-admin.conf /etc/lighttpd/conf-available2/08-pihole-admin.conf
```
```suggestion
            install -D -m 644 -T ${PI_HOLE_LOCAL_REPO}/advanced/pihole-admin.conf /etc/lighttpd/conf-available/08-pihole-admin.conf
```

PromoFaux replied:
This appears to be a typo - I can't see any mention on the internet of `conf-available2`, either :)

gstrauss replied:
You're correct.  I was looking at the Debian package for lighttpd where some of the configs in the Debian repo are in conf-available2, e.g. https://salsa.debian.org/debian/lighttpd/-/tree/master/debian/conf-available2

I have fixed this and re-pushed.

gstrauss commented:
Looking at the test failures after changing 'available2' to 'available': Did something change in the build configs?  Why is /etc/lighttpd/conf.d *not being detected* on CentOS and Fedora?  Why is /etc/lighttpd/conf-available *being detected* on CentOS and Fedora?  /etc/lighttpd/conf.d is present on my Fedora test system.

rdwebdesign commented:
Looks like tests are looking for `/etc/lighttpd/conf-available/08-pihole-admin.conf` even on CentOS and Fedora. 

https://github.com/pi-hole/pi-hole/actions/runs/3727297132/jobs/6321320558#step:6:656

rdwebdesign commented:
It is also trying to run `apt-get`.

https://github.com/pi-hole/pi-hole/actions/runs/3727297132/jobs/6321320558#step:6:630

gstrauss commented:
> Looks like tests are looking for `/etc/lighttpd/conf-available/08-pihole-admin.conf` even on CentOS and Fedora.
> 
> https://github.com/pi-hole/pi-hole/actions/runs/3727297132/jobs/6321320558#step:6:656

This is the code:
```
    # check readable /etc/lighttpd/conf*/pihole-admin.conf
    if os.path.isdir("/etc/lighttpd/conf.d"):
        check_lighttpd = test_cmd.format(
            "r", "/etc/lighttpd/conf.d/pihole-admin.conf", piholeuser
        )
        actual_rc = host.run(check_lighttpd).rc
        assert exit_status_success == actual_rc
    elif os.path.isdir("/etc/lighttpd/conf-available"):
        check_lighttpd = test_cmd.format(
            "r", "/etc/lighttpd/conf-available/08-pihole-admin.conf", piholeuser
        )
        actual_rc = host.run(check_lighttpd).rc
        assert exit_status_success == actual_rc
```
Do you see why the first condition would fail on CentOS or Fedora?  That condition test `if os.path.isdir("/etc/lighttpd/conf.d")` is expected to pass on CentOS and Fedora, and not to reach the second `elif` condition.  The second `elif` condition is expected to *fail* on CentOS and Fedora.

rdwebdesign commented:
> Do you see why the first condition would fail on CentOS or Fedora? 

No.
Unless the directory is absent on the test container.

gstrauss commented:
I think I see the issue.  The commands have to run inside the container.  The `os.path.isdir()` is running in the host running the container, which is probably Debian-based.

rdwebdesign commented:
Now it's failing everywhere... I hate this tests.

gstrauss commented:
> Now it's failing everywhere... I hate this tests.

No, I didn't get the syntax right.  I had to check for shell command success (0) not boolean truth value (1), as non-zero is an indication of shell command failure.

**The new tests I added in this PR now pass.**

rdwebdesign commented:
Yeah... I saw your code when the tests were failing. They failed because the function test was reversed (like you noticed).

gstrauss commented:
I added a comment header to pihole-admin.conf (copied from advanced/lighttpd.conf.*) and set `server.document-root = "/var/www/html"` for requests to `^/admin/`.  While /var/www/html is the default value in the Debian system-provided lighttpd.conf, that is not the case for Fedora system-provided lighttpd.conf, so specifying in the pi-hole-specific pihole-admin.conf is forward-looking to make it safe to use system-provided lighttpd.conf in new installations.

PromoFaux commented:
Tested functionality:

Debug Log shows correct header information:
```
*** [ DIAGNOSING ]: Dashboard headers
[✓] Web interface X-Header: X-Pi-hole: The Pi-hole Web interface is working!
```

IP just shows splash page:

![image](https://user-images.githubusercontent.com/1998970/208508294-912206b8-2950-4c68-9f79-ed3121edf2f4.png)

pi.hole redirects to admin:
![image](https://user-images.githubusercontent.com/1998970/208508344-61ce2fec-b399-4b0e-9cb7-79c9e3c71b10.png)
